src/config/config_type.rs

```diff
@@ -30,17 +30,42 @@ pub enum ConfigType<'a> {
 }
 
 impl ConfigType<'_> {
-    /// Deserializes the configuration to `T` based on the input type. Errors if
-    /// the data produces an invalid config.
-    pub fn deserialize<T, P>(self, filter_name: &str) -> Result<T, Error>
+    /// Deserializes takes two type arguments `Static` and `Dynamic` representing
+    /// the types of a static and dynamic configuration respectively.
+    ///
+    /// If the configuration input is a [ConfigType::Static], then it is directly
+    /// marshalled into the returned `Static` instance.
+    ///
+    /// Otherwise if the input is a [ConfigType::Dynamic] then the contained Protobuf
+    /// data is decoded into a type `Dynamic` instance, after which the decoded
+    /// value is converted into the returned `Static` instance.
+    /// As a result [TryFrom] must be implemented from the `Dynamic` to the `Static`
+    /// type.
+    ///
+    /// It returns both the marshalled, as well as, a JSON representation
+    /// of the provided config.
+    /// It returns an error if any of the serialization or deserialization steps fail.
+    pub fn deserialize<Static, Dynamic>(
+        self,
+        filter_name: &str,
+    ) -> Result<(serde_json::Value, Static), Error>
     where
-        P: prost::Message + Default,
-        T: for<'de> serde::Deserialize<'de> + TryFrom<P, Error = ConvertProtoConfigError>,
+        Dynamic: prost::Message + Default,
+        Static: serde::Serialize
+            + for<'de> serde::Deserialize<'de>
+            + TryFrom<Dynamic, Error = ConvertProtoConfigError>,
```

Comment:
This doesn't have to be in this PR, but I would like to combine these two type parameters into single trait, so that we can enforce it as part of the type system. Then you also only need to provide a single parameter.

```rust
trait FilterConfig {
   type Dynamic: prost::Message + Default;
   type Static: serde::Serialize
             + for<'de> serde::Deserialize<'de>
             + TryFrom<Self::Dynamic, Error = ConvertProtoConfigError>;
}

// src/filters/debug.rs
impl FilterConfig for Config {
    type Static = Self;
    type Dynamic = ProtoDebug;
}
```